### ISSUE TITLE
Polygonal clipping area?

### ISSUE THREAD

**Issue by briantjacobs** [openseadragon/openseadragon#1540]
I've been using the multi-image functionality to align different space images--zoom in on one area captured by one telescope, keep zooming in and show tiles from a different telescope. Some of these image have irregular edges, an area of no data. As to not occlude things i would want to mask out the no-data areas. 

Here's a sketch of this:
![mutli-clip-opaque-01](https://user-images.githubusercontent.com/786463/46089354-4a36a200-c17c-11e8-8b0c-edbdf7bdd13b.jpg)

I want to use jpeg for these tilesets to keep tile size lower, but that means I cant use png transparency. I see the `clip` parameter for tilesets but it appears that only works for a rectangle, and a rect would cut out a lot of data. Can you think of a way to apply an irregular alpha mask to a jpeg tileset?

Ideal outcome. Even more ideal, a feathered irregular mask for better blending
![mutli-clip-01](https://user-images.githubusercontent.com/786463/46089401-60dcf900-c17c-11e8-8e97-22bf53920d14.jpg)

**Comment by iangilman** [comment]
Instead of clipping them out, would you be okay with having that irregular outside shape rendered as an overlay on top of them? The end result would look the same, as long as you don't need anything to show through that area. If that would work for you, I recommend the SVG overlay plugin: 

https://github.com/openseadragon/svg-overlay

Of course supporting polygonal clipping areas would also be neat! It would be a lot more work though.

**Comment by briantjacobs** [comment]
Here's a less trivial example:
![image](https://user-images.githubusercontent.com/41201364/46168854-97437280-c267-11e8-8810-58154067ab3e.png)
The area on the left is a lower detail space tileset. The area on the right is a higher res tileset that i can continue to zoom into, aligned to the lower res, in multi-image. I'm trying to make tilesets blend together better. And a step towards that would be to remove the black area that represents no-data in the high-res image. Which is to say that I think an SVG overlay would just cover things up i'm still interested in.

**Comment by iangilman** [comment]
Yep, you'll want true clipping there. Well, are you up for making a patch to add that feature to OSD? You'd need to add a function like `setPolygonClip` to TiledImage and then use that value when drawing similar to the current clip: 

https://github.com/openseadragon/openseadragon/blob/master/src/tiledimage.js#L1910

I'm happy to answer questions as you go.

**Comment by briantjacobs** [comment]
within `drawTiles`:
```
    if (tiledImage._polyClip) {
        tiledImage._drawer.saveContext(useSketch);

        var clipPoly = tiledImage._polyClip.map(function(arr) {
            var point = tiledImage.imageToViewportCoordinates(arr[0], arr[1], true)
            point = point.rotate(-tiledImage.getRotation(true), tiledImage._getRotationPoint(true))
            var clipPoint = tiledImage._drawer.viewportCoordToDrawerCoord(point);
            if (sketchScale) {
                clipPoint = clipPoint.times(sketchScale);
            }
            if (sketchTranslate) {
                clipPoint = clipPoint.translate(sketchTranslate);
            }
            return clipPoint
        })

        tiledImage._drawer.setPolyClip(clipPoly, useSketch);

        usedClip = true;
    }
```
Within `$.Drawer`
```
    viewportCoordToDrawerCoord: function(point) {
        var topLeft = this.viewport.pixelFromPointNoRotate(point, true);
        return new $.Point(
            topLeft.x * $.pixelDensityRatio,
            topLeft.y * $.pixelDensityRatio,
        );
    },
```
```

**Comment by briantjacobs** [comment]
Sorry did not mean to close the ticket.

Given a `tiledImage` with a `polyClip` option as such, an array of x/y arrays

```
polyClip: [[83418,34],[84208,1692],[85935,996],[86712,2664],...]
```
The following functions work to do a pol yclip! Performance seems unaffected with a list of coordinates derived from a 13kb SVG I'm using.

within `drawTiles`:
```
    if (tiledImage._polyClip) {
        tiledImage._drawer.saveContext(useSketch);

        var clipPoly = tiledImage._polyClip.map(function(arr) {
            var point = tiledImage.imageToViewportCoordinates(arr[0], arr[1], true)
            point = point.rotate(-tiledImage.getRotation(true), tiledImage._getRotationPoint(true))
            var clipPoint = tiledImage._drawer.viewportCoordToDrawerCoord(point);
            if (sketchScale) {
                clipPoint = clipPoint.times(sketchScale);
            }
            if (sketchTranslate) {
                clipPoint = clipPoint.translate(sketchTranslate);
            }
            return clipPoint
        })

        tiledImage._drawer.setPolyClip(clipPoly, useSketch);

        usedClip = true;
    }
```
Within `$.Drawer`
```
    viewportCoordToDrawerCoord: function(point) {
        var topLeft = this.viewport.pixelFromPointNoRotate(point, true);
        return new $.Point(
            topLeft.x * $.pixelDensityRatio,
            topLeft.y * $.pixelDensityRatio,
        );
    },
```
```
    setPolyClip: function(poly, useSketch) {
        if (!this.useCanvas) {
            return;
        }

        var context = this._getContext( useSketch );
        context.beginPath();
        poly.forEach(function(coord,i){
          if (i == 0) {
              context.moveTo(coord.x,coord.y);
          } else {
              context.lineTo(coord.x,coord.y);
          }
        })
        context.closePath();    
        context.clip();
    },
```

It looks like there are a lot of other places to accommodate this, where `clip` is used in other functions. But putting this in here before I go away on leave for a few weeks, for the record.

**Comment by iangilman** [comment]
Looks fantastic! Thank you for sharing it here. If you're up for turning it into a patch when you return, that would be awesome :-)

**Comment by zero41120** [comment]
I was working on a proof of concept.
Currently, my project uses the [FabricJS plug-in](https://github.com/altert/OpenseadragonFabricjsOverlay) with a custom logic to load tiled custom image as an overlay because we needed cropping. That implementation is rather buggy and inefficient.

So I looked into OSD and see if we can achieve the same thing.
@briantjacobs has a working solution, but doesn't support all use cases I was looking for.
Since this issue is back to 2018, briantjacobs may have forgotten this feature.

Here is my [implementation ](https://github.com/zero41120/openseadragon/commit/2d971af4458c79c9b69a3411a1623a1f75b673a4)based on  briantjacobs' logic, with support for multiple polygons.

Here is a demo:
![jc1PJwi3CZ](https://user-images.githubusercontent.com/10471270/73981880-129b4680-48e8-11ea-80ff-4a408476d674.gif)

**Comment by iangilman** [comment]
Wonderful, thank you for taking this on!

**Comment by iangilman** [comment]
Fixed by #1772.
